src/app/Fake.DotNet.NuGet/NuGet.fs

```diff
@@ -315,40 +327,113 @@ let private pack parameters nuspecFile =
-            |> TraceSecrets.guardMessage
-            |> failwith
-    with exn when parameters.PublishTrials > 0 ->
-        publish { parameters with PublishTrials = parameters.PublishTrials - 1 }
+/// dotnet nuget push command options
+type NuGetPushParams =
+    { /// Disables buffering when pushing to an HTTP(S) server to reduce memory usage.
+      DisableBuffering: bool
+      /// The API key for the server
+      ApiKey: string option
+      /// Doesn't push symbols (even if present).
+      NoSymbols: bool
+      /// Doesn't append "api/v2/package" to the source URL.
+      NoServiceEndpoint: bool
+      /// Specifies the server URL. This option is required unless DefaultPushSource config value is set in the NuGet config file.
+      Source: string option
+      /// The API key for the symbol server.
+      SymbolApiKey: string option
+      /// Specifies the symbol server URL.
+      SymbolSource: string option
+      /// Specifies the timeout for pushing to a server.
+      Timeout: TimeSpan option
+      /// Number of times to retry pushing the package
+      PushTrials: int}
+
+    static member Create() =
+        { DisableBuffering = false
+          ApiKey = None
+          NoSymbols = false
+          NoServiceEndpoint = false
+          Source = None
+          SymbolApiKey = None
+          SymbolSource = None
+          Timeout = None
+          PushTrials = 5 }
+
+type NuGetParams with
+    member internal x.NuGetPushOptions = 
+        let normalize str = if String.isNullOrEmpty str then None else Some str
+    
+        { DisableBuffering = false
+          ApiKey = normalize x.AccessKey
+          NoSymbols = false
+          NoServiceEndpoint = false
+          Source = normalize x.PublishUrl
+          SymbolApiKey = normalize x.SymbolAccessKey
+          SymbolSource = normalize x.SymbolPublishUrl
+          Timeout = Some x.TimeOut
+          PushTrials = x.PublishTrials }
+
+    member internal x.ToolOptions = ToolOptions.Create x.ToolPath "push" x.WorkingDir true
+    member internal x.Nupkg = (x.OutputPath @@ packageFileName x |> Path.getFullName)
+
+let private toPushCliArgs param =
+    let ifTrue x b =
+        if b then Some x
+        else None
+
+    [
+        param.ApiKey
+        param.DisableBuffering |> ifTrue "-DisableBuffering"
+        param.NoSymbols |> ifTrue "-NoSymbols"
+        param.NoServiceEndpoint |> ifTrue "-NoServiceEndpoint"
+        param.Source |> Option.map (sprintf "-Source %s")
+        param.SymbolApiKey |> Option.map (sprintf "-SymbolApiKey %s")
+        param.SymbolSource |> Option.map (sprintf "-SymbolSource %s")
+        param.Timeout |> Option.map string |> Option.map (sprintf "-Timeout %s")
+    ]
+    |> List.choose id
+    |> String.concat " "
+
+module Private =
+    /// For internal use
+    let rec push (options : ToolOptions) (parameters : NuGetPushParams) (toCliArgs : NuGetPushParams -> string) nupkg =
```

Comment:
So this only currently handles `dotnet nuget push` we should think about other commands as well. 
- Do we add all operations by adding new functions one by one on demand?
- Should we use a similar model as `dotnet` uses where you have a "low"-level `DotNet.exec` function where others like `dotnet test` build on top.
- Something different?

Personally I'd lean to something along the second option, but if there are good reasons to do it differently I'm all ears. Note that I'm not proposing to implement all options in this PR, but whatever we decide to release should be extendible later.

Reply:
The second option sounds like a good idea so I'll not reuse this in the `DotNet` module, but rather use the `exec` in there instead. I'll for now make this `NuGet.Private.push` internal instead and maybe it could be generalized to be the low level one?